File: Water_Jug_Problem.py
class Water_jug_Puzzle:
    def __init__(self):
        self.four = 0  # 4 liter jug
        self.three = 0  # 3 liter jug

    def rule_1(self):
        # Fill 3 liter jug completely from faucet
        if self.three < 3:
            self.three = 3
            return True
        return False

    def rule_2(self):
        #Fill 4 liter jug completely from faucet
        if self.four < 4:
            self.four = 4
            return True
        return False

    def rule_3(self):
        # Empty 4 liter jug
        if self.four >= 0:
            self.four = 0
            return True
        return False

    def rule_4(self):
        # Empty 3 liter jug 
        if self.three >= 0:
            self.three = 0
            return True
        return False

    def rule_5(self):
        # Fill 4 liter jug from 3 liter jug 
        if self.four < 4:
            if self.four == 0:
                self.four += self.three
                self.three = 0
                return True
            else:
                self.three -= (4 - self.four)
                self.four = 4
            return True
        else:
            return False

    def rule_6(self):
        # Fill 3 liter jug from 4 liter jug 
        if self.three < 3:
            if self.three == 0:
                self.four -= self.three
                self.three = 3
                return True
            else:
                self.four -= (3 - self.three)
                self.three = 3
            return True
        else:
            return False

# Driver Code
if __name__ == '__main__':
    myJug = Water_jug_Puzzle()
    print(f'Initial State: ({myJug.four}, {myJug.three})')
    while(myJug.four != 2):
        input_rule = int(input("Enter rule: "))
        if input_rule == 0:
            break
        elif input_rule == 1:
            steps = myJug.rule_1()
        elif input_rule == 2:
            steps = myJug.rule_2()
        elif input_rule == 3:
            steps = myJug.rule_3()
        elif input_rule == 4:
            steps = myJug.rule_4()
        elif input_rule == 5:
            steps = myJug.rule_5()
        elif input_rule == 6:
            steps = myJug.rule_6()
        else:
            print('Invalid rule !')
            continue

        if not steps:
            print('Choose another rule!')
        print(f'Current State: ({myJug.four}, {myJug.three})')
print('You have solved the puzzle !')

File: WaterJugProblem.pl
% Declaring condition for final state
start(2,0):-write(' 4 liter Jug:   2 | 3 liter Jug:   0|\n'),
            write('Problem Solved!\n').      

% Starting from the initial state and take rules as input from the user
start(A,B):-write(' 4 liter Jug:   '),write(A),write('| 3 liter Jug:   '),
            write(B),write('|\n'),
            write(' Enter Rule::'),
            read(N),
            contains(A,B,N).

% Rule logic
contains(_,B,1):-start(4,B).
contains(A,_,2):-start(A,3).
contains(_,B,3):-start(0,B).
contains(A,_,4):-start(A,0).
contains(A,B,5):-N is B-4+A, start(4,N).
contains(A,B,6):-N is A-3+B, start(N,3).
contains(A,B,7):-N is A+B, start(N,0).
contains(A,B,8):-N is A+B, start(0,N).

% Main function to implement the problem and display instructions
waterjug(A,B):-write('WATER JUG PROBLEM'),nl,nl,
        write('Follow the Rules: '),nl,
        write('Rule 1: Fill 4 liter Jug\n'),nl,
        write('Rule 2: Fill 3 liter Jug\n'),nl,
        write('Rule 3: Empty 4 liter Jug\n'),nl,
        write('Rule 4: Empty 3 liter Jug\n'),nl,
        write('Rule 5: Fill 4 liter Jug with 3 liter Jug\n'),nl,
        write('Rule 6: Fill 3 liter Jug with 4 liter Jug\n'),nl,
        write('Rule 7: Empty 3 liter Jug into 4 liter Jug\n'),nl,
        write('Rule 8: Empty 4 liter Jug into 3 liter Jug\n'),nl,nl,
        write(' 4 liter Jug:   0 | 3 liter Jug:   0'),nl,
        write(' Enter Rule::'),
        read(N),nl,
        contains(0,0,N).